### PR TITLE
[#6791] test both itree options (regex vs glob) (4-2-stable)

### PR DESCRIPTION
**Pull request by d-w-moore**

### CODE REVIEW

**Comment by trel** [comment]
please add a test or two around using the pattern and ignore options together in different combinations.

separately, what happens if both regex and glob are used at the same time?  maybe those tests should be part of a different issue, not sure...

File: scripts/irods/test/test_itree.py [hunk 1]
```diff
@@ -46,6 +46,23 @@ def test_depth(self):
...
+        # Using forward and reverse combination of -P and -I options, catch objects matching /bar[0-9]/
+        self.admin.assert_icommand(["itree", "-P", "ba*", "-I", "*ar?"], "STDOUT_SINGLELINE", "Found 0 collections and 10 data objects")
+        self.admin.assert_icommand(["itree", "-I", "*ar?", "-P", "ba*"], "STDOUT_SINGLELINE", "Found 0 collections and 10 data objects")
+
+        # Using regular expressions, catch objects matching /bat[0-9]/
+        self.admin.assert_icommand(["itree", "--ignore-regex", ".*ar.", "--pattern-regex", "ba.*"], "STDOUT_SINGLELINE", "Found 0 collections and 10 data objects")
+
+        # Using regular expressions with extended syntax, catch objects matching /(bar|bat)[0-2]/x
+        self.admin.assert_icommand(["itree",  "--regex-extended", "--ignore-regex", ".*[3-9]", "--pattern-regex", "ba(r|t)[[:digit:]]"],
+                                    "STDOUT_SINGLELINE", "Found 0 collections and 6 data objects")
```
Comment (korydraughn) [inline scripts/irods/test/test_itree.py:64]: Do we have tests that verify the output contains only the matched objects?

If not, then these assertions seem like good cases to do that. You can capture the value returned by `assert_icommand` and do that. That way if anything changes in the future, we'll catch it.

Comment (d-w-moore) [reply scripts/irods/test/test_itree.py:64]: It's my observation that none of the other tests for itree do that, thus far.  Should an issue be made for that?

**Comment by d-w-moore** [comment]
> separately, what happens if both regex and glob are used at the same time? maybe those tests should be part of a different issue, not sure...

regex and glob can be used together, eg --pattern and --ignore-regex would make sense.   --pattern and --pattern-regex would not make sense, however, and possibly should be caught as an error. At present, pretty sure the later option would override the earlier one.

File: scripts/irods/test/test_itree.py (hunk 1 above)
Comment (trel) [reply scripts/irods/test/test_itree.py:64]: separate or along with this set - either is fine.

**Comment by trel** [comment]
> together would not make sense

yes, please catch anything that should result in `402000 USER_INCOMPATIBLE_PARAMS`

File: scripts/irods/test/test_itree.py [hunk 2]
```diff
@@ -46,6 +46,23 @@ def test_depth(self):
         for index,item in enumerate(directory_names):
             self.admin.assert_icommand(["itree", "-L", str(1 + index)], 'STDOUT_SINGLELINE', ["  " * (1 + index) + item])
 
+    def test_pattern_and_ignore_options_together__6791_6806(self):
+        patterns = ['foo', 'bar', 'bat']
+        for i in range(10):
+            for p in patterns:
+                self.admin.assert_icommand(['itouch', p + str(i)])
+
+        # Using forward and reverse combination of -P and -I options, catch objects matching /bar[0-9]/
```
Comment (trel) [inline scripts/irods/test/test_itree.py:55]: matches `/bat[0-9]/`, rather than `bar`, right?

File: scripts/irods/test/test_itree.py (hunk 1 above)
Comment (trel) [inline scripts/irods/test/test_itree.py:64]: can use `STDOUT_MULTILINE` to list the expected results a bit more specifically.

File: scripts/irods/test/test_itree.py [hunk 3]
```diff
@@ -46,6 +46,23 @@ def test_depth(self):
...
+        self.admin.assert_icommand(["itree", "-P", "ba*", "-I", "*ar?"], "STDOUT_SINGLELINE", "Found 0 collections and 10 data objects")
+        self.admin.assert_icommand(["itree", "-I", "*ar?", "-P", "ba*"], "STDOUT_SINGLELINE", "Found 0 collections and 10 data objects")
+
+        # Using regular expressions, catch objects matching /bat[0-9]/
+        self.admin.assert_icommand(["itree", "--ignore-regex", ".*ar.", "--pattern-regex", "ba.*"], "STDOUT_SINGLELINE", "Found 0 collections and 10 data objects")
```
Comment (trel) [inline scripts/irods/test/test_itree.py:60]: can use `STDOUT_MULTILINE` to list the expected results a bit more specifically.

File: scripts/irods/test/test_itree.py [hunk 4]
```diff
@@ -46,6 +46,23 @@ def test_depth(self):
...
 
+    def test_pattern_and_ignore_options_together__6791_6806(self):
+        patterns = ['foo', 'bar', 'bat']
+        for i in range(10):
+            for p in patterns:
+                self.admin.assert_icommand(['itouch', p + str(i)])
+
+        # Using forward and reverse combination of -P and -I options, catch objects matching /bar[0-9]/
+        self.admin.assert_icommand(["itree", "-P", "ba*", "-I", "*ar?"], "STDOUT_SINGLELINE", "Found 0 collections and 10 data objects")
+        self.admin.assert_icommand(["itree", "-I", "*ar?", "-P", "ba*"], "STDOUT_SINGLELINE", "Found 0 collections and 10 data objects")
```
Comment (trel) [inline scripts/irods/test/test_itree.py:57]: can use `STDOUT_MULTILINE` to list the expected results a bit more specifically.

File: scripts/irods/test/test_itree.py (hunk 2 above)
Comment (alanking) [reply scripts/irods/test/test_itree.py:55]: The `-I` option has a value of `*ar?`, which would suggest that `bar` would match and `bat` would not,?

File: scripts/irods/test/test_itree.py [hunk 5]
```diff
@@ -46,17 +46,41 @@ def test_depth(self):
...
             self.admin.assert_icommand(["itree", "-L", str(1 + index)], 'STDOUT_SINGLELINE', ["  " * (1 + index) + item])
 
+    def test_pattern_and_ignore_options_together__6791_6806(self):
+        patterns = ['foo', 'bar', 'bat']
+        for i in range(10):
+            for p in patterns:
+                self.admin.assert_icommand(['itouch', p + str(i)])
+        to_catch = [r'\b{}\b'.format('bat'+str(i)) for i in range(10)]
+
+        # Using forward and reverse combination of -P and -I options, catch 'baz*' objects as result
```
Comment (trel) [inline scripts/irods/test/test_itree.py:56]: catch 'bat*'

**Comment by alanking** [comment]
Seems to be one unresolved comment left here and this is probably good to go (of course, waiting on irods/irods_client_icommands#340 as well)

**Comment by alanking** [comment]
Looks like the comments have been resolved. Please squash to taste and leave off #s for now and we will have one more look.

**Review by alanking — COMMENTED** [review]
Looks good to me. Is this also related to #6806? Consider linking that one in your commit message as well, if so.

Please squash to taste, leave off #s, and we'll take one last look.

**Comment by d-w-moore** [comment]
> Looks good to me. Is this also related to #6806? Consider linking that one in your commit message as well, if so.
> 
> Please squash to taste, leave off #s, and we'll take one last look.

Related yes, in that instead of allowing multiple search patterns via repeating -I or -P, we are instead giving the user the ability to search using extended regex  syntax (`-x`), which includes the or (`|`) operator.

File: scripts/irods/test/test_itree.py [hunk 6]
```diff
@@ -46,16 +46,45 @@ def test_depth(self):
...
+        # Using forward and reverse combination of -P and -I options, catch 'bat*' objects as result
+        self.admin.assert_icommand(["itree", "-P", "ba*", "-I", "*ar?"], "STDOUT_MULTILINE", to_catch, use_regex = True)
+        self.admin.assert_icommand(["itree", "-I", "*ar?", "-P", "ba*"], "STDOUT_MULTILINE", to_catch, use_regex = True)
+
+        # Using regular expressions, catch objects matching /bat[0-9]/
+        self.admin.assert_icommand(["itree", "--ignore-regex", ".*ar.", "--pattern-regex", "ba.*"], "STDOUT_MULTILINE", to_catch, use_regex = True)
+
+        # Using regular expressions with extended syntax, catch objects matching /(bar|bat)[0-2]/x
+        to_catch = [r'\b{}\b'.format(prefix+str(i)) for i in range(3) for prefix in ('bar','bat')]
+        self.admin.assert_icommand(["itree",  "--regex-extended", "--ignore-regex", ".*[3-9]", "--pattern-regex", "ba(r|t)[[:digit:]]"],
```
Comment (alanking) [inline scripts/irods/test/test_itree.py:66]: Codacy claims... trailing whitespace...

**Comment by alanking** [comment]
Okay. I think that we should link both issues to the commit since both issue numbers are also being referenced in the name of the test.

File: scripts/irods/test/test_itree.py (hunk 6 above)
Comment (d-w-moore) [reply scripts/irods/test/test_itree.py]: whitespace removed.

**Comment by d-w-moore** [comment]
Ok - squashed with all the appropriate issue numbers applied.

**Comment by trel** [comment]
looking very nice.

**Review by korydraughn — APPROVED** [review]
Looks good to me.

Add the pound when ready.

**Review by alanking — APPROVED** [review]

**Review by alanking — APPROVED** [review]

**Pull request merged**